src/main.rs

```diff
@@ -1,3 +1,207 @@
+        let mut members_tokens_map = HashMap::new();
+
+        chat.messages.par_iter().chunks(messages_per_thread).for_each_with(sender, move |sender, messages: Vec<&Message>| {
+            let mut chunk_tokens_map: HashMap<Token<'_>, usize> = HashMap::new();
+            let mut chunk_members_tokens_map: HashMap<Person<'_>, HashMap<Token<'_>, usize>> = HashMap::new();
+            for message in messages.iter().filter(|message| message.message_type != MessageType::Service) {
+                let from = message.from.clone().unwrap();
+                for entity in message.text_entities.iter().filter(|entity| !entity.text_type.is_meta()) {
+                    for token in entity.text.split([' ', ',', '.','(', ')', '-', '!', '?', '\'', '\"', '\n', '\t']).filter(|s| !s.is_empty()) {
+                        let token = Token::from(remove_emojis(token)); // <-- such a performance hit!
+                        *chunk_tokens_map.entry(token.clone()).or_insert(0) += 1;
+                        match chunk_members_tokens_map.get_mut(&from) {
+                            Some(map) => *map.entry(token).or_insert(0) += 1,
+                            None => {
+                                let member_occurences_map = HashMap::from([(token, 1)]);
+                                chunk_members_tokens_map.insert(from.clone(), member_occurences_map);
+                            }
+                        }
+                    } 
+                }
+            }
+            sender.send((chunk_tokens_map, chunk_members_tokens_map)).unwrap();
+        });
+        for _ in 0..jobs {
+            let (chunk_tokens_map, chunk_members_tokens_map): (HashMap<Token<'_>, usize>, _) = receiver.recv().unwrap();
+            merge_maps_with(&mut tokens_map, chunk_tokens_map, |tokens_map, token, occurences| *tokens_map.entry(token).or_insert(0) += occurences);
+            for (member, map) in chunk_members_tokens_map {
+                match members_tokens_map.get_mut(&member) {
+                    None => {
+                        members_tokens_map.insert(member, map);
+                    },
+                    Some(mergee) => {
+                        merge_maps_with(mergee, map, |mergee, member, occurences| *mergee.entry(member).or_insert(0) += occurences);
+                    },
+                }
+            }
+        }
+        Self {
+            num_tokens: tokens_map.len(),
+            members_tokens_map,
+            tokens_map,
+        }
+    }
+}
+
+fn remove_emojis(string: &str) -> String {
+    use unicode_segmentation::UnicodeSegmentation;
+    let graphemes = string.graphemes(true);
+
+    let is_not_emoji = |x: &&str| emojis::get(x).is_none();
+
+    graphemes.filter(is_not_emoji).collect()
+}
+
+fn merge_maps_with<K: Eq + PartialEq + Hash, F: Fn(&mut HashMap<K, usize>, K, usize)>(dst: &mut HashMap<K, usize>, src: HashMap<K, usize>, f: F) {
+    for (key, occurences) in src {
+        f(dst, key, occurences)
+    }
 }
```

Comment:
Your code editor doesn't add a new line at the end of the file. It's quite convenient, though it's a minor-minor.